src/build.rs

```diff
@@ -347,27 +371,46 @@ fn create_disk_image(
         .write_all(&kernel_info_block)
         .with_context(|e| format!("Could not write output bootimage file: {}", e))?;
 
-    // write out kernel elf file
-    let kernel_size = kernel.metadata()?.len();
-    let mut buffer = [0u8; 1024];
-    loop {
-        let (n, interrupted) = match kernel.read(&mut buffer) {
-            Ok(0) => break,
-            Ok(n) => (n, false),
-            Err(ref e) if e.kind() == io::ErrorKind::Interrupted => (0, true),
-            Err(e) => Err(e)?,
-        };
-        if !interrupted {
-            output.write_all(&buffer[..n])?
+    fn write_file_to_file(output: &mut File, datafile: &mut File) -> Result<usize, Error> {
+        let data_size = datafile.metadata()?.len();
+        let mut buffer = [0u8; 1024];
+        let mut acc = 0;
+        loop {
+            let (n, interrupted) = match datafile.read(&mut buffer) {
+                Ok(0) => break,
+                Ok(n) => (n, false),
+                Err(ref e) if e.kind() == io::ErrorKind::Interrupted => (0, true),
+                Err(e) => Err(e)?,
+            };
+            if !interrupted {
+                acc += n;
+                output.write_all(&buffer[..n])?
+            }
         }
+
+        assert!(data_size == acc as u64);
+
+        Ok(acc)
     }
 
-    let padding_size = ((512 - (kernel_size % 512)) % 512) as usize;
-    let padding = [0u8; 512];
-    output
-        .write_all(&padding[..padding_size])
-        .with_context(|e| format!("Could not write output bootimage file: {}", e))?;
+    fn pad_file(output: &mut File, written_size: usize, padding: &[u8]) -> Result<(), Error> {
+        let padding_size = (padding.len() - (written_size % padding.len())) % padding.len();
+        output.write_all(&padding[..padding_size]).context("Could not write to output file")?;
```

Comment:
I prefer `with_context` because `context` does not show the underlying error when printed. With `with_context` we could display the reason for the failed write too (e.g. access denied, etc).